[lldb] Fix data race in statusline format handling

JDevlieghere opened the pull request:
This fixes a data race between the main thread and the default event handler thread. The statusline format option value was protected by a mutex, but it was returned as a pointer, allowing one thread to access it while another was modifying it.

Avoid the data race by returning format values by value instead of by pointer.

adrian-prantl commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -1534,15 +1534,13 @@ bool Debugger::FormatDisassemblerAddress(const FormatEntity::Entry *format,
   FormatEntity::Entry format_entry;
+  if (format)
+    format_entry = *format;
+  else if (exe_ctx != nullptr && exe_ctx->HasTargetScope())
```
```suggestion
  else if (exe_ctx && exe_ctx->HasTargetScope())
```

adrian-prantl commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -1534,15 +1534,13 @@ bool Debugger::FormatDisassemblerAddress(const FormatEntity::Entry *format,
                                          const ExecutionContext *exe_ctx,
                                          const Address *addr, Stream &s) {
   FormatEntity::Entry format_entry;
+  if (format)
```
This is a weird API.

adrian-prantl commented on lldb/source/Interpreter/OptionValue.cpp:
```diff
@@ -380,11 +380,11 @@ bool OptionValue::SetLanguageValue(lldb::LanguageType new_language) {
   if (const OptionValueFormatEntity *option_value = GetAsFormatEntity())
-    return &option_value->GetCurrentValue();
-  return nullptr;
+    return option_value->GetCurrentValue();
```
This looks safe.

adrian-prantl commented on lldb/source/Core/Statusline.cpp:
```diff
@@ -144,9 +144,9 @@ void Statusline::Redraw(bool update) {
+  FormatEntity::Format(format, stream, &symbol_ctx, &exe_ctx, nullptr, nullptr,
+                       false, false);
 
   Draw(std::string(stream.GetString()));
```
Draw(stream.GetString().str())

adrian-prantl reviewed (approved):
As far as I can tell this looks safe now.

JDevlieghere committed:
[lldb] Fix data race in statusline format handling

This fixes a data race between the main thread and the default event
handler thread. The statusline format option value was protected by a
mutex, but it was returned as a pointer, allowing one thread to access
it while another was modifying it.

Avoid the data race by returning format values by value instead of by
pointer.

JDevlieghere committed:
Formatting
